Fix #192: Handle local dependencies for Java classes

jvican opened the pull request:
Aside from fixing #192, the two first commits add more tests and guard against
the occurrence of `NoType` in the traversers.

Zinc incrementally compiles both Java and Scala sources. Most of the time,
Scala sources depend on Java sources and it's important that when Java sources
are modified, Scala sources that depend on those modifications are also
compiled again. Supporting this use case is simple -- Scalac pickles
information from Java class files and turns them into symbols that can be
handled by Zinc.

The situation starts to get trickier when it's turned around. If we want our
Java sources to be recompiled when there are Scala changes, we have to detect
the dependencies at the Java level. Java sources are compiled by javac, so we
have no symbol information whatsoever we can use to detect dependencies. To
solve this problem, Zinc class-loads the generated class files by Javac and
reflectively detects any dependency at the class level. Thanks to the clas file
[constant pool](https://docs.oracle.com/javase/specs/jvms/se7/html/jvms-4.html#jvms-4.4),
we can also know which classes a given class file refers to -- buying us the
analog of the name hashing algorithm implemented by @gkossakowski in Scala
sources.


Before, local class files were correctly honoured by Zinc and linked to their
original source file (remember that javac creates a new class file for every
class that is not top level). However, local classes were not analyzed by
`ClassAPI` and registered as dependencies.

These commits introduce local dependencies by inheritance from Java classes to
Scala classes. It applies a similar fix to the one provided in https://github.com/gkossakowski/sbt/commit/47416b394e3730d67367da9f011f8ef3de23b576.
In the commit messages, I dive into a more succint explanation of the fixes,
feel free to check them out.

This work is done by the [Scala Center](https://scala.epfl.ch/).

jvican committed:
Add checks against NoType

They are not relevant and their effect on performance is minimal, but
it's good to have them for consistency. We do check for `NoSymbol` as
well as `null` when it comes to symbols. It feels weird if we don't for
types.

jvican committed:
Add complementary tests to local-class-inheritance

Add more tests aside from local-class-inheritance to check that
declaring anonymous clases and local classes in Scala that depend on
Java artifacts works (even though the anonymous class is desugared into
a local class by Scala).

This just gives us a wider safety net, the only test that was checking
this key functionality was 'local-class-inheritance'.

jvican committed:
Add partial support to local classes in Java

Java local, inner and anonymous classes are not handled correctly by the
current algorithm. When they extend Scala classes and those are changed,
Zinc does not recompile the dependent Java code. See
https://github.com/sbt/zinc/issues/192.

These changes introduce the first step to support recompilation of Scala
sources when Java classes depend on them via inheritance. Before, even if we
had member references to Scala class files, the invalidation logic would
not recompile Java sources. There needs to be an inheritance relation
for that to happen.

The approach taken in this commit fixes this issue, but in a non-ideal
way. Currently, the dependency is tracked in the enclosing class of the
local class as an inheritance dependency. This overcompiles sources in
several scenarios -- if the Java enclosing class is extended by other
Java classes, those will be recompiled as well because the invalidation
logic assigns this contract to inheritance dependencies.

What we want to express -- a local inheritance dependency -- was added in
https://github.com/gkossakowski/sbt/commit/47416b394e3730d67367da9f011f8ef3de23b576.
Local dependencies are special cased by the invalidation logic so that
subclasses of the enclosing class are not recompiled.

For now, we stay with this solution. In the upcoming commit, this normal
inheritance dependency will be turned into local and some of the new
logic in `Analyze` will be removed, `productsToSource` already maps
local class files to the sources of their enclosing classes.

jvican committed:
Fix #192: Handle local dependencies for Java classes

The previous change added a global inheritance relationship between the
enclosing class of the local Java classes and the Scala sources. As we
discussed in https://github.com/jvican/zinc/commit/f022f91ec81303dd054fd3de589ca3ad8136b507,
this is far from ideal since Zinc overcompiles subclasses of the
enclosing class.

This commit turns the previous inheritance dependencies into local
dependencies, and therefore compile only the enclosing class but not
their subclasses.

As you see, classes that cannot be mapped to a source are supposed to be
either local classes (that are therefore mapped to the source of their
enclosing class) or stale classes that are present in the output
directory. In the latter case, we ignore such classes and not process
them.

jvican commented:
Is there any reason why CI information is not available? Is the CI running?

lightbend-cla-validator commented:
Hi @jvican,

Thank you for your contribution! We really value the time you've taken to put this together.

Before we proceed with reviewing this pull request, please sign the Lightbend Contributors License Agreement:

[http://www.lightbend.com/contribute/cla](http://www.lightbend.com/contribute/cla)

Duhemm commented on internal/zinc-classfile/src/main/scala/sbt/internal/inc/classfile/Analyze.scala:
```diff
@@ -121,21 +119,28 @@ private[sbt] object Analyze {
+        localClassesToSources.contains(cls.getName))
+      val localInherited: Map[String, Set[String]] =
+        readInheritanceDependencies(localClasses)
+      nonLocalInherited foreach {
```
Shouldn't it be `localInherited` here?

jvican replied:
Yes, it should, thanks for the catch.

romanowski commented on internal/zinc-apiinfo/src/main/scala/sbt/internal/inc/ClassToAPI.scala:
```diff
@@ -20,7 +20,7 @@ object ClassToAPI {
     {
       val pkgs = packages(c).map(p => new api.Package(p))
       val cmap = emptyClassMap
-      val defs = c.filter(isTopLevel).flatMap(toDefinitions(cmap))
```
can we rename `c` into e.g. `classes`?

jvican replied:
Sure.

romanowski commented on internal/zinc-classfile/src/main/scala/sbt/internal/inc/classfile/Analyze.scala:
```diff
@@ -63,22 +63,34 @@ private[sbt] object Analyze {
+      // Map local classes to the sources of their enclosing classes
+      val localClassesToSources = localClassesOrStale.flatMap { cls =>
+        val enclosingLoadedCls = Option(cls.getEnclosingClass)
+        val sourceOfEnclosing = enclosingLoadedCls.map(binaryToSourceName(_))
```
Maybe:

```scala
val localClassesToSources: Map[String, String] =  (for {
  cls <- localClassesOrStale
  enclosingLoadedCls <- Option(cls.getEnclosingClass)
  sourceOfEnclosing <- binaryToSourceName(enclosingLoadedCls)
} yield (cls.getName, sourceOfEnclosing))(collection.breakout)
```

jvican replied:
👍

romanowski commented on internal/zinc-classfile/src/main/scala/sbt/internal/inc/classfile/Analyze.scala:
```diff
@@ -63,22 +63,34 @@ private[sbt] object Analyze {
+          sourceName <- binaryToSourceName(loadedClass)
+        } yield sourceName
+
+        if (nonLocalSourceName.isDefined) nonLocalSourceName
```
Maybe: 
```scala
nonLocalSourceName orElse localClassesToSources.get(className)
```

jvican replied:
👍

romanowski commented on internal/zinc-classfile/src/main/scala/sbt/internal/inc/classfile/Analyze.scala:
```diff
@@ -63,22 +63,34 @@ private[sbt] object Analyze {
+        val fromClassName: String = {
+          getMappedSource(fromBinaryName) match {
+            case Some(sourceName) => sourceName
+            case None             => return // It could be a stale class file
```
TBH I don't like returns in scala code (but I don't want force my way here :) ). Maybe instead:
```scala
def processDependency(onBinaryName: String, context: DependencyContext, fromBinaryName: String): Unit = 
 getMappedSource(fromBinaryName) match {
  case None => 
  case Some(fromClassName) =>
    // Old code here
}
```

jvican replied:
The return was already there, I didn't introduce it. I didn't want to create major diff change and tried to keep it minimal. I can rewrite it if you like.

romanowski replied:
It's your call, I understand your point here.

romanowski commented:
LGTM with few small code style suggestions

jvican committed:
Address feedback and rewrite `processDependency`

This commit rewrites a part of `processDependency` to remove the
`return` clause and use become clearer to read. It also addresses
@romanowski's suggestions.

jvican committed:
Rename `c` into classes

This commit is done separately in case maintainers prefer to remove it.
It renames a variables that was poorly named but also changes the git
history. I think this change is justified but I would understand
arguments against it, since it's only an aesthetic change.

eed3si9n commented:
LGTM
Thanks @jvican and Scala Center for the contribution.

jvican commented:
I will open another PR (perhaps a batch of other PRs) to backport all my changes to 2.10 sources. I've just realized my changes only affected 2.11 sources.

romanowski commented:
I think we can wait a while and backport changes in one PR (e.g. my speedup was also targeted to 2.11/2.12 only).

jvican commented:
Yes, that's what I had in mind -- it's why I said batch of other PRs.
